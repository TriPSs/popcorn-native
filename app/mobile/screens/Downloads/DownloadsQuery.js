import gql from 'graphql-tag'

export const DownloadsQuery = gql`
  query downloads($offset: Float!) {
    downloads(limit: 25, offset: $offset) {
      _id
      __typename
      quality
      status
      movie {
        _id
        title
        synopsis
        type
        download{
          downloadQuality
          downloadStatus
          downloading
        }
        images {
          poster {
            thumb
            high
          }
        }
      }
      episode {
        _id
        title
        number
        synopsis
        type
        download{ 
          downloadQuality
          downloadStatus
          downloading
        }
        show {
          title
          images {
            poster {
              thumb
              high
            }
          }
        }
      }
    }
  }
`
