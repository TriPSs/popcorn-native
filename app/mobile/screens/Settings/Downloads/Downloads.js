import React from 'react'
import PropTypes from 'prop-types'
import { View } from 'react-native'

import withDownloadManager from 'modules/DownloadManager/withDownloadManager'
import dimensions from 'modules/dimensions'
import i18n from 'modules/i18n'

import Container from 'components/Container'
import Typography from 'components/Typography'

import Download from './Download'

export const styles = {

  root: {
    paddingTop: dimensions.UNIT * 2,
  },

  title: {
    marginLeft: dimensions.UNIT,
    marginRight: dimensions.UNIT,
    marginBottom: dimensions.UNIT / 2,
  },

  container: {
    paddingTop: dimensions.UNIT / 2,
    paddingBottom: dimensions.UNIT,

    display: 'flex',
    flexDirection: 'row',
  },

  noDownloads: {
    padding: dimensions.UNIT,
  },

}

export const Downloads = ({ executeQuery, data, downloadManager, loading }) => (
  <Container style={styles.root}>

    <Typography
      style={styles.title}
      variant={'headline6'}>
      {i18n.t('Active Downloads')}
    </Typography>

    {(!data?.activeDownloads || data?.activeDownloads?.length === 0) && (
      <Container
        style={[styles.container, styles.noDownloads]}
        elevation={2}>

        <View>
          {loading && (
            <Typography variant={'subtitle2'}>
              {i18n.t('Loading')}
            </Typography>
          )}

          {!loading && (
            <Typography variant={'subtitle2'}>
              {i18n.t('No active downloads')}
            </Typography>
          )}
        </View>
      </Container>
    )}

    {data?.activeDownloads?.map((download) => (
      <Download
        key={download._id}
        download={download}
        downloadManager={downloadManager}
        refreshScreen={executeQuery}
      />
    ))}

  </Container>
)

Downloads.propTypes = {
  executeQuery: PropTypes.func.isRequired,
  data: PropTypes.object,
}

Downloads.defaultProps = {
  data: null,
}

export default withDownloadManager(Downloads)
