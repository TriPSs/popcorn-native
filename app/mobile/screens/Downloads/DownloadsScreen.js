import React, { useEffect, useRef } from 'react'
import { StyleSheet, View, StatusBar, FlatList, InteractionManager, RefreshControl } from 'react-native'
import { useLazyQuery } from '@apollo/react-hooks'

import i18n from 'modules/i18n'
import colors from 'modules/colors'
import dimensions from 'modules/dimensions'
import withDownloadManager from 'modules/DownloadManager/withDownloadManager'
import fetchMoreUpdateQuery from 'modules/GraphQL/helpers/fetchMoreUpdateQuery'

import Typography from 'components/Typography'
import ItemInfo from 'mobile/components/ItemInfo'

import { DownloadsQuery } from './DownloadsQuery'

const styles = StyleSheet.create({

  root: {
    flex: 1,
    backgroundColor: colors.BACKGROUND,
  },

  container: {
    marginTop: dimensions.UNIT / 2,
    paddingBottom: dimensions.UNIT * 2,
  },

  noResults: {
    marginTop: dimensions.UNIT * 2,
    textAlign: 'center',
  },
})

export const DownloadsScreen = ({ navigation, downloadManager }) => {
  const flatListRef = useRef(null)
  const [executeQuery, { data, loading, fetchMore }] = useLazyQuery(
    DownloadsQuery,
    {
      variables: {
        offset: 0,
      },
    },
  )

  useEffect(() => {
    // Execute the query after the component is done navigation
    InteractionManager.runAfterInteractions(() => {
      // Execute the query
      executeQuery()
    })
  }, [])

  const items = !data
    ? []
    : data.downloads

  const renderCard = ({ item }) => {
    const empty = item.loading

    return (
      <ItemInfo
        empty={empty}
        item={item.episode || item.movie}
        variant={'playable'}
        quality={item.quality}
        status={item.status}
        truncateSynopsis
      />
    )
  }

  const renderNothingFound = () => (
    <Typography
      style={styles.noResults}
      variant={'subtitle1'}>
      {i18n.t('No results found')}
    </Typography>
  )

  return (
    <View style={styles.root}>

      <StatusBar
        backgroundColor={colors.STATUS_BAR_TRANSPARENT}
        animated />

      <FlatList
        removeClippedSubviews
        ref={flatListRef}
        refreshControl={
          <RefreshControl
            refreshing={items.length > 0 && loading}
            onRefresh={() => executeQuery()}
            colors={[colors.PRIMARY_COLOR_200]}
            progressBackgroundColor={colors.BACKGROUND_TABS}
          />
        }
        data={items.length === 0
          ? Array(8).fill({ loading: true })
          : items
        }
        scrollEnabled={items.length > 0}
        numColumns={1}
        initialNumToRender={8}
        windowSize={8}
        renderItem={renderCard}
        ListHeaderComponent={() => (
          <View style={{
            marginTop: dimensions.STATUSBAR_HEIGHT,
          }} />
        )}
        ListEmptyComponent={renderNothingFound}
        ListFooterComponent={() => <View style={{ width: dimensions.UNIT * 2 }} />}
        keyExtractor={(item, index) => item
          ? `${item._id}-${index}`
          : `${index}`
        }
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={false}
        onEndReached={items.length > 0
          ? fetchMoreUpdateQuery('downloads', data, fetchMore)
          : null
        }
        onEndReachedThreshold={dimensions.CARD_HEIGHT_SMALL * 4}
        contentContainerStyle={styles.container}
      />

    </View>
  )
}

DownloadsScreen.propTypes = {}

export default withDownloadManager(DownloadsScreen)
