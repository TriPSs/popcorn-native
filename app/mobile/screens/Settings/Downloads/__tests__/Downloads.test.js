import React from 'react'
import { shallow } from 'enzyme'

import { Downloads } from '../Downloads'

describe('(Screens) Settings > Downloads', () => {

  const downloadManager = {
    removeDownload: jest.fn(),
    onPress: jest.fn(),
  }

  it('should render empty list correctly', () => {
    const result = shallow(
      <Downloads
        downloadManager={downloadManager}
        executeQuery={jest.fn()}
        data={null}
      />,
    )

    expect(result).toMatchSnapshot()
  })

  it('should render list correctly', () => {
    const result = shallow(
      <Downloads
        downloadManager={downloadManager}
        executeQuery={jest.fn()}
        data={{ activeDownloads: [{ _id: 1 }] }}
      />,
    )

    expect(result).toMatchSnapshot()
  })


})
