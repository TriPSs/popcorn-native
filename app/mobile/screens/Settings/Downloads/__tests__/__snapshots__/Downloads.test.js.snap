// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Screens) Settings > Downloads should render empty list correctly 1`] = `
<Container
  elevation={0}
  style={
    Object {
      "paddingTop": 16,
    }
  }
>
  <Typography
    color="white"
    component={null}
    emphasis="high"
    fontWeight="default"
    style={
      Object {
        "marginBottom": 4,
        "marginLeft": 8,
        "marginRight": 8,
      }
    }
    textProps={Object {}}
    variant="headline6"
  >
    Active Downloads
  </Typography>
  <Container
    elevation={2}
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingTop": 4,
        },
        Object {
          "padding": 8,
        },
      ]
    }
  >
    <View>
      <Typography
        color="white"
        component={null}
        emphasis="high"
        fontWeight="default"
        style={null}
        textProps={Object {}}
        variant="subtitle2"
      >
        No active downloads
      </Typography>
    </View>
  </Container>
</Container>
`;

exports[`(Screens) Settings > Downloads should render list correctly 1`] = `
<Container
  elevation={0}
  style={
    Object {
      "paddingTop": 16,
    }
  }
>
  <Typography
    color="white"
    component={null}
    emphasis="high"
    fontWeight="default"
    style={
      Object {
        "marginBottom": 4,
        "marginLeft": 8,
        "marginRight": 8,
      }
    }
    textProps={Object {}}
    variant="headline6"
  >
    Active Downloads
  </Typography>
  <Download
    data={null}
    download={
      Object {
        "_id": 1,
      }
    }
    downloadManager={
      Object {
        "onPress": [MockFunction],
        "removeDownload": [MockFunction],
      }
    }
    key="1"
    refreshScreen={[MockFunction]}
  />
</Container>
`;
